### ISSUE TITLE
Purpose of [[TemplateMap]] in Realm Record.

### ISSUE THREAD

**Issue by dSalieri** [tc39/ecma262#2742]
You know if you look at the definition of realm you will see following:
> Before it is evaluated, all ECMAScript code must be associated with a [realm](https://tc39.es/ecma262/#realm). Conceptually, a realm consists of a set of intrinsic objects, an ECMAScript global environment, all of the ECMAScript code that is loaded within the scope of that global environment, and other associated state and resources.
>
> A [realm](https://tc39.es/ecma262/#realm) is represented in this specification as a Realm Record with the fields specified in [Table 27](https://tc39.es/ecma262/#table-realm-record-fields)

That's all.

Well, there is no definitions about special fields. But if we look at the table 27 we will see something...

Since, I am interested in `[[TemplateMap]]`, I will ask about it. There is `"meaning"` in the table, but table don't tell you about her purpose, it says about structure of `[[TemplateMap]]`, that's it, and there are no somewhere words about purpose of `[[TemplateMap]]`. Hence you don't understand purpose, if you're not maintainer of ecma262. So my question is "What is `[[TemplateMap]]` actual, why does the specification need the `[[TemplateMap]]`?"

P.S `[[TemplateMap]]` is the only thing you can't figure out by looking at the spec in the realm. Would like to see an explanation not only here but also in the specification.

**Comment by Jack-Works** [comment]
It is some magic behavior of the tagged template strings.

https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-gettemplateobject

**Comment by Jack-Works** [comment]
![image](https://user-images.githubusercontent.com/5390719/163791906-98ec57ce-98af-481a-a06b-9b75ffc569e4.png)

**Comment by dSalieri** [comment]
@Jack-Works sure, I saw that link that you showed. Your example is very interesting, but this is not still answer to my question about purpose and needing of `[[TemplateMap]]`. I want to understand sense of this concept.

**Comment by bathos** [comment]
It allows caching results (of e.g. parsers of other languages) in association with the template literal parse node. Template tag functions can perform work related to the literal segments just once, with subsequent evaluations only needing to perform work related to the substitution values. This is what makes them "templates".

**Comment by dSalieri** [comment]
@bathos In other words, `[[TemplateMap]]` caches parts of template literals as the arguments so that when the same syntax node is called again, the template literals is not parsed. That is, it is an optimization. So? Is it a micro-optimization or redundant, what do you think?

I'm asking this because I'm trying to understand the expediency of such a decision. I see that `[[TemplateMap]]` was created back in ES6 and this concept has not changed. Really it is enough bottleneck where productivity seriously falls?

**Comment by ljharb** [comment]
It enables template tag functions to cache results without having to create a key based on every argument, so yes.

**Comment by bathos** [comment]
> That is, it is an optimization.

It can be, but it's affording an _identity_ associated with the parse node and this enables more than just performance optimizations. It is a kind of language extension point.

**Comment by devsnek** [comment]
concretely it allows you to write something like

```js
const cache = new Map();
function myCoolTag(strings, ...values) {
  if (!cache.has(strings)) {
    cache.set(strings, computeComplexLogic(strings));
  }
  return cache.get(strings)(values);
}

myCoolTag`whatever ${1} stuff`;
```

the identify of the array `strings` is always the same for each spec parse node, so you can very efficiently parametrize the `values`.

**Comment by dSalieri** [comment]
@devsnek I am afraid, that your example won't work because each new execution of expression 
```JavaScript
myCoolTag`whatever ${1} stuff`;
```
will create new record in a Map.

Look at the folowing code:
```JavaScript
function tag(){
    return arguments[0];
}
tag`123` === tag`123`; /// false
```
but
```JavaScript
function tag(){
    return arguments[0];
}
function tagA(){
    return tag`123`
}
tagA() === tagA(); /// true
```
So it should be the same parse Node, otherwise you won't see optimization.

**Comment by devsnek** [comment]
@dSalieri you can think of it as the source location, here's an example: https://engine262.js.org/#gist=74ea9624bb9f61a9b1c42f7c26ce9417

**Comment by dSalieri** [comment]
> @dSalieriвы можете думать об этом как об исходном местоположении, вот пример: https://engine262.js.org/#gist=74ea9624bb9f61a9b1c42f7c26ce9417

That's it what I meant :)

**Comment by dSalieri** [comment]
> @dSalieri you can think of it as the source location, here's an example: https://engine262.js.org/#gist=74ea9624bb9f61a9b1c42f7c26ce9417

That's it what I meant :)

**Comment by bathos** [comment]
@dSalieri It sounds like you’ve got the idea. The expression in the example devsnek gave is the source text “with identity” preserved and exposed to ES code as the identity of the frozen array. The same array will be received by the tag function each time a specific tagged template expression (i.e., at some particular source text location) is evaluated, which becomes apparent if such an expression appears within a function called multiple times or within the body of for-loop, etc.

Initially these arrays were keyed by the specific combination of string values, not the parse nodes, which led to problems (e.g. for GC); this was also less rather than more powerful in a strict sense because key-by-strings can be polyfilled, but key-by-parse-node is a unique capability afforded only by this syntax. Nonetheless, the change [did lead to problems]((https://webreflection.medium.com/a-tiny-disastrous-ecmascript-change-fadc05c83e69) as it was made after the original semantics had been around a while. The author of that article, @webreflection, has written a number of other articles about template tags and their usage and has studied the problem space deeply and authored several libraries that would be instructive. I believe that taken together, those materials are a great resource to explore next if you’re looking to understand how it's leveraged in practice more.

Earlier I suggested checking out lit-html as an example. Here is [a tiny demo](https://hollow-fabulous-triangle.glitch.me/) I made just now with the goal of helping to illustrate which aspects of it are likely to be interesting. In the “updates log” that appears there, note that only the very first one “creates the template” (parsing HTML, etc), while the first two “stamp the template” (because there are two instances keyed like parse loc → container) . All subsequent calls only update the parts that are changing in both instances.

**Comment by dSalieri** [comment]
@bathos you've written:
> Recreating those elements on every evaluation wouldn’t just be slower — it wouldn’t achieve the same thing at all.

Hmm. How do we get different results with the same literal strings but different parse nodes?

**Comment by dSalieri** [comment]
@bathos sorry, I should have made my question more precise. Let's do it again.
You wrote:
> Recreating those elements on every evaluation wouldn’t just be slower — it wouldn’t achieve the same thing at all.

I asked the question:
> How do we get different results with the same template literals but different parse nodes?

Under that I mean following:
```JavaScript
function tag(){
    return arguments[0];
}
tag`123` === tag`123`; /// false
```
How can I get different results by equal template literals? Your quote: `it wouldn’t achieve the same thing at all`.

**Comment by bathos** [comment]
You don't. The template literals there are the same. In the HTML example, that's a key part of the recipe. I think I see what might be throwing you off though as the example I gave creates two "instances". The instances aren't keyed by just the template array, though; they have a "composite" key, a template array + a host node. the 'template' created the first time has a weakmap like host->live instance. Roughly. I'm eliding some details and only mean to get the gist across as this is just one particular example/pattern. (Although it's a solid illustration, there's a risk of overemphasizing aspects of this which are unique to the needs of HTML templating.)

**Comment by dSalieri** [comment]
@bathos excuse me for repeating myself, but could you demonstrate your words?
> Recreating those elements on every evaluation wouldn’t just be slower — it wouldn’t achieve the same thing at all.

Could I look at this practically? I want to see confirmation of your words.

**Comment by bathos** [comment]
I provided a glitch demo earlier that illustrates interpolated-values-only updates where the static portions of the template string (corresponding to an HTML element in the example) are likewise static, i.e. their identity is preserved across every update. If you poke around in the console there maybe it’ll be more evident what I meant? For example you could run a CSS animation on them: if the elements were being recreated every time it updated instead of preserved, then the animation would reset every time, too. Perhaps I’m misunderstanding your question, not sure.

**Comment by dSalieri** [comment]
@bathos yeah, I saw your demonstration that based on caching the same parse node. The thing is what I asked to show visa versa behaviour of template literals when recreating elements wouldn't achieve the same thing
> if the elements were being recreated/replaced every time it updated, then the animation would also reset every second instead of continuing to run

Well, it is not necessary to recreate the element, it is enough to mutate it. Your example does just that.

I think I kind of understand you.
